irohad/consensus/yac/impl/yac.cpp

```diff
@@ -24,336 +24,349 @@ auto &getRound(const std::vector<iroha::consensus::yac::VoteMessage> &state) {
+        ++it;
       }
-
-      Yac::Yac(YacVoteStorage vote_storage,
-               std::shared_ptr<YacNetwork> network,
-               std::shared_ptr<YacCryptoProvider> crypto,
-               std::shared_ptr<Timer> timer,
-               ClusterOrdering order,
-               Round round,
-               rxcpp::observe_on_one_worker worker,
-               logger::LoggerPtr log)
-          : log_(std::move(log)),
-            cluster_order_(order),
-            round_(round),
-            worker_(worker),
-            notifier_(worker_, notifier_lifetime_),
-            vote_storage_(std::move(vote_storage)),
-            network_(std::move(network)),
-            crypto_(std::move(crypto)),
-            timer_(std::move(timer)) {}
-
-      Yac::~Yac() {
-        notifier_lifetime_.unsubscribe();
+      future_states_.erase(future_states_.begin(), it);
+    }
+    if (not state.empty()) {
+      onState(std::vector<VoteMessage>{std::make_move_iterator(state.begin()),
+                                       std::make_move_iterator(state.end())});
+    }
+    auto vote = crypto_->getVote(hash);
+    // TODO 10.06.2018 andrei: IR-1407 move YAC propagation strategy to a
+    // separate entity
+    votingStep(vote);
+  }
+
+  rxcpp::observable<Answer> Yac::onOutcome() {
+    return notifier_.get_observable();
+  }
+
+  // ------|Network notifications|------
+
+  template <typename T, typename P>
+  void removeMatching(std::vector<T> &target, const P &predicate) {
+    target.erase(std::remove_if(target.begin(), target.end(), predicate),
+                 target.end());
+  }
+
+  template <typename CollectionType, typename ElementType>
+  bool contains(const CollectionType &haystack, const ElementType &needle) {
+    return std::find(haystack.begin(), haystack.end(), needle)
+        != haystack.end();
+  }
+
+  /// moves the votes not present in known_keys from votes to return value
+  void Yac::removeUnknownPeersVotes(std::vector<VoteMessage> &votes,
+                                    ClusterOrdering &order) {
+    auto known_keys =
+        order.getPeers() | boost::adaptors::transformed([](const auto &peer) {
+          return peer->pubkey();
+        });
+    removeMatching(
+        votes, [known_keys = std::move(known_keys), this](VoteMessage &vote) {
+          if (not contains(known_keys, vote.signature->publicKey())) {
+            log_->warn("Got a vote from an unknown peer: {}", vote);
+            return true;
+          }
+          return false;
+        });
+  }
+
+  void Yac::onState(std::vector<VoteMessage> state) {
+    std::unique_lock<std::mutex> guard(mutex_);
+    auto &proposal_round = getRound(state);
+
+    if (not crypto_->verify(state)) {
+      log_->warn("Crypto verification failed for message. Votes: [{}]",
+                 boost::algorithm::join(
+                     state | boost::adaptors::transformed([](const auto &v) {
+                       return v.signature->toString();
```

Comment:
`v.signature->toString()` size is about 64 or 128 symbols? What is the result string length for state of size 100?

How about redusing hash size?
```suggestion
                       return v.signature->toString().substr(0,7)+"..";
```